Home Task #1_1

Sovianum opened the pull request:
1. Refactored url addition syntax;
2. Fixed server_lib.

Commit:
1. Refactored url addition syntax;
2. Fixed server_lib.

Antonboom commented on server_lib.js:
```diff
@@ -129,51 +137,98 @@ exports.callbackTemplates = {
 
-exports.getStaticServer = function(_urlCallbackDict) {
-    let urlCallbackDict = _urlCallbackDict || {};
+class URLMap {
```
Го назовем Router?
Можно кстати в экспорт вынести getRouter() { return new Router(); }

Antonboom replied:
Сделано

Commit:
Renamed URLMap to Router

Antonboom commented on server_lib.js:
```diff
@@ -129,51 +137,98 @@ exports.callbackTemplates = {
-                callback(request, response);
-                return
+        for (let url in urlCallbackDict) {
+            if (_isRegExpStr(url) && request.url.match(new RegExp(_regexFromRegexStr(url)))) {
```
Как насчет чего-нибудь такого?
```javascript
let callback_url = Object.keys(a).find((url) => { return _isRegExpStr(url) && request.url.match(_getRegexFromRegexStr(url)) });
```
Не проверял, но идея понятна

Sovianum replied:
Идея интересна. Я попробую переделать на что-то подобное, когда заменю префиксы на несколько словарей, чтобы несколько раз одно и то же не рефакторить.

Antonboom replied:
Ок, я тебе написал в трелле в доработку.
Плюс, встроенные методы вроде как быстрее работают, да и функциональный стиль приятнее выглядит, чем императивный, особенно в отношении JS

Sovianum replied:
Переделал

Antonboom commented on server_lib.js:
```diff
@@ -129,51 +137,98 @@ exports.callbackTemplates = {
+        if (callback) {
+            callback(request, response);
+            console.log("Complete match succeeded: " + request.url);
+            return
```
else после return - лишний блок. Можно оставить, если нужно выделить ветку else.
Там дальше много такого

Sovianum replied:
Убрал лишние блоки

Antonboom commented on main.js:
```diff
@@ -1,27 +1,22 @@
-    "/signup/": new lib.BindedFileFunc("./static/html/signup.html"),
-    "/": new lib.BindedFileFunc("./static/html/index.html"),
+let router = new lib.Router();
+router.addPlainURL("/index/", new lib.BindedFile("./static/html/index.html"));
```
Разве так не красивее?)
```javascript
const PLAIN_URLS = ['index', 'about', 'game', 'leaders', 'login', 'singin', 'signup'];
PLAIN_URLS.forEach((url) => { 
    router.addPlainUrl(`/${url}`, new lib.BindFile(`./static/html/${url}.html`);
});
```

Sovianum replied:
Заменил.

Antonboom commented on server_lib.js:
```diff
@@ -129,51 +137,98 @@ exports.callbackTemplates = {
+module.exports.getRegexStr = getRegexStr;
+module.exports.getPrefixStr = getPrefixStr;
+module.exports.Router = Router;
+module.exports.getStaticServer = getStaticServer;
```
В конце файла пустая строка.

Sovianum replied:
Сделал

Antonboom commented on server_lib.js:
```diff
@@ -1,46 +1,53 @@
+const fs = require("fs");
+const http = require("http");
+const path = require("path");
+
```
Выбери единый стиль отступов. Обычно между глобальными объявлениями - две строки, внутри функции - одна.

Sovianum replied:
Сделал

Antonboom commented on server_lib.js:
```diff
@@ -129,51 +137,98 @@ exports.callbackTemplates = {
+    }
+}
+
+function getStaticServer (_URLMap) {
```
URL_MAP -> Router 
И для параметров нет понятия модификатора доступа - незачем начинать с "_". :)

Sovianum replied:
Сделал

Antonboom commented on server_lib.js:
```diff
@@ -1,46 +1,53 @@
+const REG_EXP_PREFIX = "REGEX";
+const PREFIX_STR_PREFIX = "PREFIX";
+
+function FileByPrefixFunc (request, response) {
```
Это где-то используется? Если да, то как работает?
И снова Func в названии.

Sovianum replied:
Нет, осталась от предыдущей версии. Выпилил.

Antonboom commented on server_lib.js:
```diff
@@ -1,46 +1,53 @@
-exports.BindedFileFunc = function(pagePath) {
+function BindedFile (pagePath) {
     return function (request, response) {
         console.log("Requested URL: " + request.url);
```
Можно ставить запятую вместо, а не соединять "+" - console.log сделает это за тебя.

Sovianum replied:
Заменил

Antonboom commented on server_lib.js:
```diff
@@ -1,46 +1,53 @@
                 response.end();
             } else {
                 response.write(data);
+                response.end();
```
У тебя ниже response.end() уже есть :)
Это касается и блока выше

Sovianum replied:
Поправил здесь и еще в одном месте

Antonboom commented on server_lib.js:
```diff
@@ -129,51 +137,98 @@ exports.callbackTemplates = {
+
+        for (let url in urlCallbackDict) {
+            if (_isPrefixStr(url) && request.url.startsWith(_prefixFromPrefixStr(url))) {
+                let prefix = _prefixFromPrefixStr(url);
```
prefix можно вынести повыше - тогда не придется два раза вызывать _prefixFromPrefixStr(url) - меньше объектов.

Sovianum replied:
То, что объектов будет меньше - вопрос открытый, так как если не выполняется первое условие, то не будет ни одного вызова. Но prefix до if красивее, согласен

Antonboom replied:
Подловил :(

Sovianum replied:
Исправил

Antonboom commented on server_lib.js:
```diff
@@ -49,7 +56,7 @@ exports.BindedFileFunc = function(pagePath) {
-exports.FolderFileFunc = function (folderPath, urlPrefix) {
+function FolderFile (folderPath, urlPrefix) {
     return function (request, response) {
         console.log("Requested URL: " + request.url);
```
Насчет кода ниже - может быть случай, когда не будет urlPrefix??
Если нет, то может сразу делать замену префикса на folderPath?

Sovianum replied:
Может, если ты мапишь папку в /

Antonboom replied:
Да, я тоже об этом подумал.
Но, тогда разве сюда не пробросится urlPrefix - "/" ?

Sovianum replied:
В данном случае да. Но такого урла может не быть.

Antonboom replied:
Приведи пример мапинга, пожалуйста

Sovianum replied:
. => /

Antonboom replied:
Ну )
Проброситься урлпрефикс слэш, который зайдет в if(urlPrefix) - нет?
Чтобы таких вопросов не возникало и пишутся тесты.

Sovianum replied:
Убрал эту проверку

Commit:
Cosmetic repair

Antonboom commented on server_lib.js:
```diff
@@ -129,51 +137,98 @@ exports.callbackTemplates = {
+            console.log("Regex match failed: " + request.url);
+        }
+
+        let urlPair = {
```
Напиши что-ли комментарий, чтобы было более очевидно, зачем это.
Например
```javascript
/*
    For fix some interesting case:
    Dirs:    
        dir_1/mid_dir/file, 
        dir_2/file
    Routing: 
        /static/ => dir_1,  
        /static/mid_dir => dir_2
    Url:     
        /static/mid_dir/file

    Return dir_1/mid_dir/file (instead of dir_2/mid_dir/file) 
    if you do not check the length of the prefix!
*/
```

Sovianum replied:
Ок. Сделаю с утра

Sovianum replied:
Добавил

Antonboom commented on main.js:
```diff
@@ -1,27 +1,22 @@
-};
-urlDict[lib.getRegexStr("/static_js/")] = new lib.FolderFileFunc("./static/js/", "/static_js/"); // particular folder is binded to regexp
-urlDict[lib.getPrefixStr("/static/")] = new lib.FolderFileFunc("./static/", "/static/");    // particular folder is binded to prefix
+router.addRegexURL("\.*\.js", new lib.FolderFile("."));
```
main.js мы не подключаем в файлах, а все остальное упрется в корневую директорию.
Может все-таки лучше /static/js? Чтобы в файлах писать просто /file.js

Sovianum replied:
Да, это без проблем. Строку с регуляркой я сделал просто чтобы показать, что бывают урлы с регулярками)

Antonboom replied:
Нет, я про папку /static/js
Регулярку оставь - это круто!

Sovianum replied:
Тогда, я думаю, стоит для симметрии /static/css и /static/html тоже прописать.

Sovianum replied:
Я пока оставлю и то, и то, чтобы не переправлять пути в верстке.

Antonboom replied:
Ок, если что, мы найдем человека, который переправит нам пути :)

Sovianum replied:
А, нет, это у меня уже крыша едет. Чтобы отдавать по пути /file.js /static/js не подходит: все равно префикс совпадать должен.
Вместо этого тогда нужно просто регулярки оставить.

Antonboom replied:
```javascript
router.addRegexURL("\.*\.js", new lib.FolderFile("/static/js/"));
```
работать не будет?

Antonboom commented on server_lib.js:
```diff
@@ -49,25 +46,25 @@ exports.BindedFileFunc = function(pagePath) {
  * @constructor
  */
-exports.FolderFileFunc = function (folderPath, urlPrefix) {
+function FolderFile (folderPath, urlPrefix) {
```
Го тогда BindFolder, соблюдаем единообразие в названиях методов интерфейса.
Или не ок?

Sovianum replied:
Нет, здесь я полностью согласен

Commit:
FolderFile => BindedFolder

Antonboom commented on server_lib.js:
```diff
@@ -221,7 +221,7 @@ function getStaticServer (router) {
 
 module.exports.BindedFile = BindedFile;
-module.exports.FolderFile = FolderFile;
+module.exports.FolderFile = BindedFolder;
```
В main.js поменял? :)

Sovianum replied:
Да

Commit:
Some make-up with urls

Antonboom commented on server_lib.js:
```diff
@@ -1,46 +1,43 @@
+const fs = require("fs");
+
+
+const http = require("http");
```
Это можно и нужно писать рядом)
Разделяй на логические блоки тоже с умом)
```
= require
= require
= require


const
const
const


function


function
```

И можно и одним отступом, напомню, что я лишь предлагал выбрать удобный для тебя способ и рассказал, как бывает.

Sovianum replied:
На самом деле мне самому так привычнее, потому что на питоне учился)

Commit:
1. Wrapped fs calls with promises;
2. Rewritten url check from imperative to functional style;
3. Optimized out url prefixes

Antonboom reviewed (approved):
Выглядеть стало гораздо приятнее!
Молодец!

Commit:
Merge branch 'master' into dev
